kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyFrontendHandler.java

```diff
@@ -553,4 +583,52 @@ public void channelActive(ChannelHandlerContext ctx) throws Exception {
     public String toString() {
         return "KafkaProxyFrontendHandler{inbound = " + inboundCtx.channel() + ", state = " + state + "}";
     }
+
+    private void outboundChannelUsable() {
+        if (state != State.CONNECTED && state != State.NEGOTIATING_TLS) {
+            throw illegalState(null);
+        }
+        LOGGER.trace("{}: outboundChannelActive", inboundCtx.channel().id());
+        // connection is complete, so first forward the buffered message
+        for (Object bufferedMsg : bufferedMsgs) {
+            forwardOutbound(outboundCtx, bufferedMsg);
+        }
+        bufferedMsgs = null; // don't pin in memory once we no longer need it
+        if (pendingReadComplete) {
+            pendingReadComplete = false;
+            channelReadComplete(outboundCtx);
+        }
+        state = State.OUTBOUND_ACTIVE;
+
+        var inboundChannel = this.inboundCtx.channel();
+        // once buffered message has been forwarded we enable auto-read to start accepting further messages
+        inboundChannel.config().setAutoRead(true);
+    }
+
+    private void onUpstreamSslContext(SslContext sslContext) {
+        this.state = State.NEGOTIATING_TLS;
+        final HostPort remote = (HostPort) outboundCtx.channel().attr(REMOTE_PEER_KEY).get();
+        final SslHandler handler = sslContext.newHandler(outboundCtx.channel().alloc(), remote.host(), remote.port());
```

Comment:
I prefer the channel pipeline being built in one place.  You have moved it here so you can get the handshakeFuture() from the sslHandler.  You could just lookup the handler from the pipeline by name `ssl`.   That saves you need to stash the REMOTE_PEER_KEY too.

Reply:
Ohh yikes that's a much better idea, thanks.